src/Compilers/CSharp/Portable/Symbols/MemberSymbolExtensions.cs

```diff
@@ -496,5 +496,30 @@ internal static string GetMemberCallerName(this Symbol member)
                 member.IsExplicitInterfaceImplementation() ? ExplicitInterfaceHelpers.GetMemberNameWithoutInterfaceName(member.Name) :
                 member.Name;
         }
+
+        /// <summary>
+        /// Locations[0] on lambda symbols covers the entire syntax, which is inconvenient but remains for compatibility.
+        /// For better diagnostics quality, use the NarrowLocation instead, which points to the "delegate" or the "=>".
+        /// </summary>
+        internal static Location NarrowLocation(this MethodSymbol methodSymbol)
```

Comment:
Is there a reason for making this an extension method? We could add a virtual internal method to a MethodSymbol with implementation:
```
return Locations[0];
```

Then we could override this method for a LambdaSymbol with special behavior, thus avoiding conditional code around symbol type/kind checks and an introduction of a new property just for the sake of being able to write code here:
```
internal SyntaxNode Syntax => _syntax;
```